File: python/examples/native/twilio/twilio_factory.py
#!/usr/bin/env python3
"""
============================================================================
TWILIO TOOLS - FACTORY PATTERN EXAMPLE
============================================================================

PATTERN: SDK Factory Pattern
─────────────────────────────────────────────────────────────────────────
Direct tool execution via MatimoInstance - the simplest way to use tools.

Use this pattern when:
✅ Building simple scripts or CLI tools
✅ Direct API calls without abstraction
✅ Quick prototyping
✅ One-off tool execution

SETUP:
─────────────────────────────────────────────────────────────────────────
1. Create .env file in project root:
   TWILIO_ACCOUNT_SID=ACxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
   TWILIO_AUTH_TOKEN=your_auth_token_here
   TWILIO_FROM_NUMBER=+15557122661
   TWILIO_TO_NUMBER=+15558675310

2. Get Twilio credentials:
   - Go to: https://console.twilio.com
   - Account SID and Auth Token are on the dashboard
   - Buy a phone number or use your trial number as TWILIO_FROM_NUMBER
   - In trial mode, TWILIO_TO_NUMBER must be a verified caller ID
   - Matimo automatically handles base64 encoding — no extra steps required!

USAGE:
─────────────────────────────────────────────────────────────────────────
  make twilio-factory

AVAILABLE TOOLS:
─────────────────────────────────────────────────────────────────────────
1. twilio-send-sms
   Parameters: account_sid (required), to (required), from (required), body (required)
   Optional: status_callback
   Returns: Message resource with sid and status: queued

2. twilio-send-mms
   Parameters: account_sid (required), to (required), from (required), media_url (required)
   Optional: body, status_callback
   Returns: Message resource with sid and num_media: "1"

3. twilio-get-message
   Parameters: account_sid (required), message_sid (required)
   Returns: Full Message resource with current status

4. twilio-list-messages
   Parameters: account_sid (required)
   Optional: to, from, date_sent, page_size
   Returns: Paginated list of Message resources

============================================================================
"""

import asyncio
import os
import sys
from datetime import datetime
from pathlib import Path
from typing import Optional

from dotenv import load_dotenv

try:
    from matimo_twilio import get_tools_path
    from matimo import Matimo
except ImportError:
    from matimo import Matimo
    get_tools_path = None

load_dotenv(Path(__file__).parent.parent.parent / ".env")


async def run_twilio_factory_examples() -> None:
    """Run factory pattern examples for Twilio messaging tools."""
    print("\n╔════════════════════════════════════════════════════════╗")
    print("║     Twilio Tools - Factory Pattern                     ║")
    print("║     (Direct execution - simplest approach)             ║")
    print("╚════════════════════════════════════════════════════════╝\n")

    # Validate required environment variables
    account_sid = os.environ.get("TWILIO_ACCOUNT_SID")
    if not account_sid:
        print("❌ Error: TWILIO_ACCOUNT_SID not set in .env")
        print('   Set it: export TWILIO_ACCOUNT_SID="ACxxxxxxxxxx"')
        print("   Get it from: https://console.twilio.com")
        sys.exit(1)

    auth_token = os.environ.get("TWILIO_AUTH_TOKEN")
    if not auth_token:
        print("❌ Error: TWILIO_AUTH_TOKEN not set in .env")
        print("   Set it: export TWILIO_AUTH_TOKEN=\"your_auth_token\"")
        print("   Get it from: https://console.twilio.com")
        sys.exit(1)

    from_number = os.environ.get("TWILIO_FROM_NUMBER")
    if not from_number:
        print("❌ Error: TWILIO_FROM_NUMBER not set in .env")
        print("   Set it: export TWILIO_FROM_NUMBER=\"+15557122661\"")
        print("   (Your Twilio phone number in E.164 format)")
        sys.exit(1)

    to_number = os.environ.get("TWILIO_TO_NUMBER")
    if not to_number:
        print("❌ Error: TWILIO_TO_NUMBER not set in .env")
        print("   Set it: export TWILIO_TO_NUMBER=\"+15558675310\"")
        print("   (In trial mode, must be a verified caller ID)")
        sys.exit(1)

    print("🔑 Account SID: Configured.")
    print(f"📤 From: {from_number}")
    print(f"📥 To:   {to_number}\n")

    # Initialize Matimo with auto-discovery to find all @matimo/* packages
    print("🚀 Initializing Matimo...")
    matimo = await Matimo.init(get_tools_path() if get_tools_path else None, auto_discover=True)

    all_tools = matimo.list_tools()
    twilio_tools = [t for t in all_tools if t.name.startswith("twilio-")]
    print(f"✅ Loaded {len(all_tools)} tools total")
    print(f"🔧 Found {len(twilio_tools)} Twilio tools: {', '.join([t.name for t in twilio_tools])}\n")

    print("════════════════════════════════════════════════════════════\n")
    print("Running Examples:")
    print("════════════════════════════════════════════════════════════\n")

    sent_message_sid: Optional[str] = None

    try:
        # ── EXAMPLE 1: List recent messages ──────────────────────────────────
        print("1️⃣   Listing recent messages...")
        list_result = await matimo.execute(
            "twilio-list-messages",
            {"account_sid": account_sid, "page_size": 5}
        )
        list_data = (list_result or {}).get("data", list_result) or {}

        if isinstance(list_data.get("messages"), list):
            messages = list_data["messages"]
            print(f"   ✅ Found {len(messages)} recent message(s)")
            for idx, msg in enumerate(messages[:3], 1):
                body_preview = (msg.get("body") or "")[:40]
                print(f"      {idx}. {msg.get('sid')} — {msg.get('direction')} — {msg.get('status')} — \"{body_preview}\"")
        else:
            print(f"   ℹ️  No messages found or unexpected response: {str(list_data)[:100]}")
        print()

        # ── EXAMPLE 2: Send SMS ───────────────────────────────────────────────
        print("2️⃣   Sending SMS message...")
        try:
            sms_result = await matimo.execute(
                "twilio-send-sms",
                {
                    "account_sid": account_sid,
                    "to": to_number,
                    "from": from_number,
                    "body": f"Hello from Matimo! Factory pattern test at {datetime.now().isoformat()}",
                }
            )
            sms_data = (sms_result or {}).get("data", sms_result) or {}

            if sms_data.get("sid"):
                sent_message_sid = sms_data["sid"]
                print("   ✅ SMS queued successfully!")
                print(f"      SID:    {sms_data['sid']}")
                print(f"      Status: {sms_data.get('status')}")
                print(f"      To:     {sms_data.get('to')}")
                print(f"      From:   {sms_data.get('from')}")
            else:
                print(f"   ❌ SMS failed: {str(sms_data)[:100]}")
        except Exception as sms_err:
            msg = str(sms_err) if isinstance(sms_err, Exception) else str(sms_err)
            print(f"   ⚠️   SMS skipped: {msg}")
            if from_number == to_number:
                print("   ℹ️  Tip: TWILIO_FROM_NUMBER and TWILIO_TO_NUMBER must be different numbers.")
                print("          In trial mode, TWILIO_TO_NUMBER must be a verified caller ID.")
        print()

        # ── EXAMPLE 3: Fetch the sent message ──────────────────────────────────
        if sent_message_sid:
            print("3️⃣   Fetching the sent message...")
            try:
                fetch_result = await matimo.execute(
                    "twilio-get-message",
                    {"account_sid": account_sid, "message_sid": sent_message_sid}
                )
                fetch_data = (fetch_result or {}).get("data", fetch_result) or {}

                if fetch_data.get("sid"):
                    print("   ✅ Message fetched!")
                    print(f"      SID:          {fetch_data['sid']}")
                    print(f"      Status:       {fetch_data.get('status')}")
                    print(f"      Direction:    {fetch_data.get('direction')}")
                    print(f"      Num Segments: {fetch_data.get('num_segments')}")
                    print(f"      Date Created: {fetch_data.get('date_created')}")
                else:
                    print(f"   ❌ Failed to fetch message: {str(fetch_data)[:100]}")
            except Exception as fetch_err:
                print(f"   ⚠️   Fetch skipped: {str(fetch_err)}")
            print()
        else:
            print("3️⃣   Fetching sent message — skipped (no SMS was queued)")
            print()

        # ── EXAMPLE 4: List messages filtered by recipient ────────────────────
        print("4️⃣   Listing messages filtered by recipient...")
        filtered_result = await matimo.execute(
            "twilio-list-messages",
            {"account_sid": account_sid, "to": to_number, "page_size": 5}
        )
        filtered_data = (filtered_result or {}).get("data", filtered_result) or {}

        if isinstance(filtered_data.get("messages"), list):
            print(f"   ✅ Found {len(filtered_data['messages'])} message(s) sent to {to_number}")
        else:
            print(f"   ℹ️  Response: {str(filtered_data)[:100]}")
        print()

    except Exception as error:
        print(f"❌ Error: {str(error)}")
        sys.exit(1)

    print("════════════════════════════════════════════════════════════")
    print("✨ Factory Pattern Example Complete!")
    print("════════════════════════════════════════════════════════════\n")


def main() -> None:
    """Run the async main function."""
    asyncio.run(run_twilio_factory_examples())


if __name__ == "__main__":
    main()

File: typescript/examples/tools/twilio/twilio-factory.ts
#!/usr/bin/env node
/**
 * ============================================================================
 * TWILIO TOOLS - FACTORY PATTERN EXAMPLE
 * ============================================================================
 *
 * PATTERN: SDK Factory Pattern
 * ─────────────────────────────────────────────────────────────────────────
 * Direct tool execution via MatimoInstance - the simplest way to use tools.
 *
 * Use this pattern when:
 * ✅ Building simple scripts or CLI tools
 * ✅ Direct API calls without abstraction
 * ✅ Quick prototyping
 * ✅ One-off tool execution
 *
 * SETUP:
 * ─────────────────────────────────────────────────────────────────────────
 * 1. Create .env file in project root:
 *    TWILIO_ACCOUNT_SID=ACxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
 *    TWILIO_AUTH_TOKEN=your_auth_token_here
 *    TWILIO_FROM_NUMBER=+15557122661
 *    TWILIO_TO_NUMBER=+15558675310
 *
 * 2. Get Twilio credentials:
 *    - Go to: https://console.twilio.com
 *    - Account SID and Auth Token are on the dashboard
 *    - Buy a phone number or use your trial number as TWILIO_FROM_NUMBER
 *    - In trial mode, TWILIO_TO_NUMBER must be a verified caller ID
 *    - Matimo automatically handles base64 encoding — no extra steps required!
 *
 * USAGE:
 * ─────────────────────────────────────────────────────────────────────────
 *   pnpm twilio:factory
 *
 * AVAILABLE TOOLS:
 * ─────────────────────────────────────────────────────────────────────────
 * 1. twilio-send-sms
 *    Parameters: account_sid (required), to (required), from (required), body (required)
 *    Optional: status_callback
 *    Returns: Message resource with sid and status: queued
 *
 * 2. twilio-send-mms
 *    Parameters: account_sid (required), to (required), from (required), media_url (required)
 *    Optional: body, status_callback
 *    Returns: Message resource with sid and num_media: "1"
 *
 * 3. twilio-get-message
 *    Parameters: account_sid (required), message_sid (required)
 *    Returns: Full Message resource with current status
 *
 * 4. twilio-list-messages
 *    Parameters: account_sid (required)
 *    Optional: to, from, date_sent, page_size
 *    Returns: Paginated list of Message resources
 *
 * ============================================================================
 */

import 'dotenv/config';
import { MatimoInstance } from 'matimo';

/**
 * Run factory pattern examples for Twilio messaging tools
 */
async function runTwilioFactoryExamples() {
  console.info('\n╔════════════════════════════════════════════════════════╗');
  console.info('║     Twilio Tools - Factory Pattern                     ║');
  console.info('║     (Direct execution - simplest approach)             ║');
  console.info('╚════════════════════════════════════════════════════════╝\n');

  // Validate required environment variables
  const accountSid = process.env.TWILIO_ACCOUNT_SID;
  if (!accountSid) {
    console.error('❌ Error: TWILIO_ACCOUNT_SID not set in .env');
    console.info('   Set it: export TWILIO_ACCOUNT_SID="ACxxxxxxxxxx"');
    console.info('   Get it from: https://console.twilio.com');
    process.exit(1);
  }

  const authToken = process.env.TWILIO_AUTH_TOKEN;
  if (!authToken) {
    console.error('❌ Error: TWILIO_AUTH_TOKEN not set in .env');
    console.info('   Set it: export TWILIO_AUTH_TOKEN="your_auth_token"');
    console.info('   Get it from: https://console.twilio.com');
    process.exit(1);
  }

  const fromNumber = process.env.TWILIO_FROM_NUMBER;
  if (!fromNumber) {
    console.error('❌ Error: TWILIO_FROM_NUMBER not set in .env');
    console.info('   Set it: export TWILIO_FROM_NUMBER="+15557122661"');
    console.info('   (Your Twilio phone number in E.164 format)');
    process.exit(1);
  }

  const toNumber = process.env.TWILIO_TO_NUMBER;
  if (!toNumber) {
    console.error('❌ Error: TWILIO_TO_NUMBER not set in .env');
    console.info('   Set it: export TWILIO_TO_NUMBER="+15558675310"');
    console.info('   (In trial mode, must be a verified caller ID)');
    process.exit(1);
  }

  console.info(`🔑 Account SID: Configured.`);
  console.info(`📤 From: ${fromNumber}`);
  console.info(`📥 To:   ${toNumber}\n`);

  // Initialize Matimo with auto-discovery to find all @matimo/* packages
  console.info('🚀 Initializing Matimo...');
  const matimo = await MatimoInstance.init({ autoDiscover: true });

  const allTools = matimo.listTools();
  const twilioTools = allTools.filter((t) => t.name.startsWith('twilio-'));
  console.info(`✅ Loaded ${allTools.length} tools total`);
  console.info(
    `🔧 Found ${twilioTools.length} Twilio tools: ${twilioTools.map((t) => t.name).join(', ')}\n`
  );

  console.info('════════════════════════════════════════════════════════════\n');
  console.info('Running Examples:');
  console.info('════════════════════════════════════════════════════════════\n');

  try {
    // ── EXAMPLE 1: List recent messages ──────────────────────────────────
    console.info('1️⃣  Listing recent messages...');
    const listResult = await matimo.execute('twilio-list-messages', {
      account_sid: accountSid,
      page_size: 5,
    });
    const listData = (listResult as any).data || listResult;

    if (listData.messages && Array.isArray(listData.messages)) {
      const messages = listData.messages;
      console.info(`   ✅ Found ${messages.length} recent message(s)`);
      messages.slice(0, 3).forEach((msg: any, idx: number) => {
        console.info(
          `      ${idx + 1}. ${msg.sid} — ${msg.direction} — ${msg.status} — "${msg.body?.substring(0, 40)}"`
        );
      });
    } else {
      console.info(
        `   ℹ️  No messages found or unexpected response: ${JSON.stringify(listData).substring(0, 100)}`
      );
    }
    console.info();

    // ── EXAMPLE 2: Send SMS ───────────────────────────────────────────────
    console.info('2️⃣  Sending SMS message...');
    let sentMessageSid: string | undefined;
    try {
      const smsResult = await matimo.execute('twilio-send-sms', {
        account_sid: accountSid,
        to: toNumber,
        from: fromNumber,
        body: `Hello from Matimo! Factory pattern test at ${new Date().toISOString()}`,
      });
      const smsData = (smsResult as any).data || smsResult;

      if (smsData.sid) {
        sentMessageSid = smsData.sid;
        console.info(`   ✅ SMS queued successfully!`);
        console.info(`      SID:    ${smsData.sid}`);
        console.info(`      Status: ${smsData.status}`);
        console.info(`      To:     ${smsData.to}`);
        console.info(`      From:   ${smsData.from}`);
      } else {
        console.info(`   ❌ SMS failed: ${JSON.stringify(smsData).substring(0, 100)}`);
      }
    } catch (smsErr: any) {
      const msg = smsErr?.message || String(smsErr);
      console.info(`   ⚠️  SMS skipped: ${msg}`);
      if (fromNumber === toNumber) {
        console.info(
          `   ℹ️  Tip: TWILIO_FROM_NUMBER and TWILIO_TO_NUMBER must be different numbers.`
        );
        console.info(`          In trial mode, TWILIO_TO_NUMBER must be a verified caller ID.`);
      }
    }
    console.info();

    // ── EXAMPLE 3: Fetch the sent message ──────────────────────────────────
    if (sentMessageSid) {
      console.info('3️⃣  Fetching the sent message...');
      try {
        const fetchResult = await matimo.execute('twilio-get-message', {
          account_sid: accountSid,
          message_sid: sentMessageSid,
        });
        const fetchData = (fetchResult as any).data || fetchResult;

        if (fetchData.sid) {
          console.info(`   ✅ Message fetched!`);
          console.info(`      SID:          ${fetchData.sid}`);
          console.info(`      Status:       ${fetchData.status}`);
          console.info(`      Direction:    ${fetchData.direction}`);
          console.info(`      Num Segments: ${fetchData.num_segments}`);
          console.info(`      Date Created: ${fetchData.date_created}`);
        } else {
          console.info(
            `   ❌ Failed to fetch message: ${JSON.stringify(fetchData).substring(0, 100)}`
          );
        }
      } catch (fetchErr: any) {
        console.info(`   ⚠️  Fetch skipped: ${fetchErr?.message || fetchErr}`);
      }
      console.info();
    } else {
      console.info('3️⃣  Fetching sent message — skipped (no SMS was queued)');
      console.info();
    }

    // ── EXAMPLE 4: List messages filtered by recipient ────────────────────
    console.info('4️⃣  Listing messages filtered by recipient...');
    const filteredResult = await matimo.execute('twilio-list-messages', {
      account_sid: accountSid,
      to: toNumber,
      page_size: 5,
    });
    const filteredData = (filteredResult as any).data || filteredResult;

    if (filteredData.messages && Array.isArray(filteredData.messages)) {
      console.info(`   ✅ Found ${filteredData.messages.length} message(s) sent to ${toNumber}`);
    } else {
      console.info(`   ℹ️  Response: ${JSON.stringify(filteredData).substring(0, 100)}`);
    }
    console.info();
  } catch (error) {
    console.error('❌ Error:', error instanceof Error ? error.message : String(error));
    process.exit(1);
  }

  console.info('════════════════════════════════════════════════════════════');
  console.info('✨ Factory Pattern Example Complete!');
  console.info('════════════════════════════════════════════════════════════\n');
}

runTwilioFactoryExamples().catch(console.error);
